File: src/convert.py
from collections import defaultdict, deque
import re
from slots import ParsedWidget

def sort_variables(variable_contents, variable_dependencies):
    # Build the adjacency list and in-degree count for the dependency graph
    adjacency_list = defaultdict(list)
    in_degree = defaultdict(int)

    # Initialize graph with all variables
    for var in variable_contents:
        in_degree[var] = 0  # Ensure all variables have an entry

    for var, dependencies in variable_dependencies.items():
        for dep in dependencies:
            adjacency_list[dep].append(var)
            in_degree[var] += 1

    # Collect all nodes with zero in-degree
    queue = deque([var for var, degree in in_degree.items() if degree == 0])
    sorted_vars = []

    # Perform topological sort
    while queue:
        current = queue.popleft()
        sorted_vars.append(current)

        # Reduce the in-degree of each neighbor
        for neighbor in adjacency_list[current]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    # Check for cycles (unsorted items)
    if len(sorted_vars) != len(variable_contents):
        raise ValueError("Cycle detected in dependencies, sorting not possible.")

    # Return the sorted variable contents
    return {var: variable_contents[var] for var in sorted_vars}

def parse_props(content: str, indent: int) -> dict:
    props = {}

    for line in content.splitlines():
        if not re.match(fr"{indent * ' '}\w", line):
            continue

        part = line[indent:]
        if part.startswith("Begin Object"):
            continue

        if "=" in part:
            key, value = part.split("=", 1)
            props[key.strip()] = value.strip()

    return props

def parse_widgets(content: str, indent: int) -> list[ParsedWidget]:
    regex = fr"{indent * ' '}Begin Object.*?\n{indent * ' '}End Object"

    objects = []

    for match in re.finditer(regex, content, re.DOTALL | re.MULTILINE):
        content = match.group(0)
        first_line = content.splitlines()[0]

        props_raw = first_line.split("Begin Object ")[1].split(" ")
        
        props = { prop.strip().split("=")[0]: prop.strip().split("=")[1] for prop in props_raw }

        children: list[ParsedWidget] = parse_widgets(content, indent + 3)
        props.update(parse_props(content, indent + 3))

        obj: ParsedWidget = {
            "props": props,
            "children": children
        }

        objects.append(obj)

    return objects

def replace_file(file: str, content: str, from_key: str, to_key: str) -> None:
    with open(file, "r", encoding="utf-8") as f:
        lines = f.read().splitlines()

    start_line = -1
    end_line = -1

    for i, line in enumerate(lines):
        if from_key in line:
            start_line = i + 1
        elif to_key in line:
            end_line = i

    # Start line shouldn't ever be -1, because we check for it in the main function
    if start_line == -1 or end_line == -1:
        print("End key not found")
        return
    
    pre = "\n".join(lines[:start_line])
    post = "\n".join(lines[end_line:])

    with open(file, "w", encoding="utf-8") as f:
        f.write(pre + "\n" + content + post)

File: src/convert.ts
import { ParsedWidget } from "./types";

export function sortVariables(
    variableContents: Map<string, string>,
    variableDependencies: Map<string, string[]>
): Map<string, string> {
    const adjacencyList = new Map<string, string[]>();
    const inDegree = new Map<string, number>();

    for (const v of variableContents.keys()) {
        inDegree.set(v, 0);
    }

    for (const [v, deps] of variableDependencies) {
        for (const dep of deps) {
            if (!adjacencyList.has(dep)) adjacencyList.set(dep, []);
            adjacencyList.get(dep)!.push(v);
            inDegree.set(v, (inDegree.get(v) ?? 0) + 1);
        }
    }

    const queue: string[] = [];
    for (const [v, degree] of inDegree) {
        if (degree === 0) queue.push(v);
    }

    const sorted: string[] = [];
    while (queue.length > 0) {
        const current = queue.shift()!;
        sorted.push(current);

        for (const neighbor of adjacencyList.get(current) ?? []) {
            const newDeg = inDegree.get(neighbor)! - 1;
            inDegree.set(neighbor, newDeg);
            if (newDeg === 0) queue.push(neighbor);
        }
    }

    if (sorted.length !== variableContents.size) {
        throw new Error("Cycle detected in dependencies, sorting not possible.");
    }

    const result = new Map<string, string>();
    for (const v of sorted) {
        result.set(v, variableContents.get(v)!);
    }
    return result;
}

export function parseProps(content: string, indent: number): Record<string, string> {
    const props: Record<string, string> = {};
    const indentStr = " ".repeat(indent);

    for (const line of content.split("\n")) {
        if (!line.startsWith(indentStr) || line.length <= indent) continue;
        if (line[indent] === " ") continue;

        const part = line.substring(indent);
        if (part.startsWith("Begin Object") || part.startsWith("End Object")) continue;

        const eqPos = part.indexOf("=");
        if (eqPos !== -1) {
            const key = part.substring(0, eqPos).trim();
            const value = part.substring(eqPos + 1).trim();
            props[key] = value;
        }
    }

    return props;
}

export function parseHeaderProps(headerLine: string): Record<string, string> {
    const props: Record<string, string> = {};
    const start = headerLine.indexOf("Begin Object ");
    if (start === -1) return props;

    const rest = headerLine.substring(start + "Begin Object ".length);
    let idx = 0;

    while (idx < rest.length) {
        while (idx < rest.length && rest[idx] === " ") idx++;
        if (idx >= rest.length) break;

        const eqPos = rest.indexOf("=", idx);
        if (eqPos === -1) break;

        const key = rest.substring(idx, eqPos).trim();
        idx = eqPos + 1;

        if (idx < rest.length && rest[idx] === '"') {
            idx++;
            const endQuote = rest.indexOf('"', idx);
            if (endQuote === -1) {
                props[key] = `"${rest.substring(idx)}"`;
                idx = rest.length;
            } else {
                props[key] = `"${rest.substring(idx, endQuote)}"`;
                idx = endQuote + 1;
            }
        } else {
            const spacePos = rest.indexOf(" ", idx);
            if (spacePos === -1) {
                props[key] = rest.substring(idx);
                idx = rest.length;
            } else {
                props[key] = rest.substring(idx, spacePos);
                idx = spacePos;
            }
        }
    }

    return props;
}

function findMatchingEnd(lines: string[], startIdx: number, indent: number): number {
    const indentStr = " ".repeat(indent);
    let depth = 0;

    for (let idx = startIdx; idx < lines.length; idx++) {
        const line = lines[idx];
        if (line.startsWith(indentStr) && line.length > indent && line[indent] !== " ") {
            const stripped = line.substring(indent);
            if (stripped.startsWith("Begin Object")) {
                depth++;
            } else if (stripped.startsWith("End Object")) {
                depth--;
                if (depth === 0) return idx;
            }
        }
    }

    return -1;
}

export function parseWidgets(content: string, indent: number): ParsedWidget[] {
    const objects: ParsedWidget[] = [];
    const lines = content.split("\n");
    const indentStr = " ".repeat(indent);

    let idx = 0;
    while (idx < lines.length) {
        const line = lines[idx];

        if (
            line.startsWith(indentStr) &&
            line.length > indent &&
            line[indent] !== " " &&
            line.substring(indent).startsWith("Begin Object")
        ) {
            const endIdx = findMatchingEnd(lines, idx, indent);
            if (endIdx === -1) {
                idx++;
                continue;
            }

            const blockLines = lines.slice(idx, endIdx + 1);
            const blockContent = blockLines.join("\n");

            const props = parseHeaderProps(line);
            const children = parseWidgets(blockContent, indent + 3);

            Object.assign(props, parseProps(blockContent, indent + 3));

            objects.push({ props, children });
            idx = endIdx + 1;
        } else {
            idx++;
        }
    }

    return objects;
}
